### PR TITLE
Add C# 9 tests for Implement* rules

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Related to #3668
Reproduces #3734

New tests are mostly copy/paste from regular test file.

### CODE REVIEW

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/ImplementISerializableCorrectly.CSharp9.cs [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+        public SerializableRecord()
+        { /*do something*/ }
+        protected SerializableRecord(SerializationInfo info, StreamingContext context)
+        { /*do something*/ }
+        public virtual void GetObjectData(SerializationInfo info, StreamingContext context)
+        { /*do something*/ }
+    }
+
+    [Serializable]
+    public class SerializableClass : ISerializable
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/ImplementISerializableCorrectly.CSharp9.cs:18]: why do you need this?

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/ImplementISerializableCorrectly.CSharp9.cs [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+    public partial record Partial_SerializableDerived_Not_CallingBase_GetObjectData_Record
+    // FN, Should be Non-compliant
+    {
+        public override partial void GetObjectData(SerializationInfo info, StreamingContext context)
+        //FN, should be second location with Invoke 'base.GetObjectData(SerializationInfo, StreamingContext)' in this method.}}
+        { /*do something*/ }
+    }
+
+    [Serializable]
+    public partial class Partial_SerializableDerived_CallingBase_GetObjectData_Class : SerializableClass
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/ImplementISerializableCorrectly.CSharp9.cs:253]: make this a record? same for `Partial_SerializableDerived_CallingBase_GetObjectData_Class` and `Partial_SerializableDerived_Not_CallingBase_GetObjectData_Class`?

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/ImplementISerializableCorrectly.CSharp9.cs [hunk 3]
```diff
@@ -0,0 +1,282 @@
+﻿using System;
+using System.Runtime.Serialization;
+
+namespace Tests.Diagnostics
+{
+    [Serializable]
+    public record SerializableRecord : ISerializable
+    {
+        public SerializableRecord()
+        { /*do something*/ }
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/ImplementISerializableCorrectly.CSharp9.cs:10]: I know it's a nitpick, but by removing the `/*do something*/` comment and by bringing the brackets on the same line with the method declaration, you'd remove some clutter from the test cases

```
public SerializableRecord() {}
protected SerializableRecord(SerializationInfo info, StreamingContext context) {}
public virtual void GetObjectData(SerializationInfo info, StreamingContext context) {}
```
is much better

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/ImplementSerializationMethodsCorrectly.CSharp9.cs [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        void OnDeserializing(StreamingContext context) // FN, attribute should not be on a local method
+        { }
+
+        [OnDeserialized()]
+        void OnDeserialized(StreamingContext context) // FN, attribute should not be on a local method
+        { }
+
+        [OnDeserialized()]
+        static void OnDeserializedStatic(StreamingContext context) // FN, attribute should not be on a local nor static method
+        { }
```
Comment (andrei-epure-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/ImplementSerializationMethodsCorrectly.CSharp9.cs:75]: you can now have `private partial void` and `private partial int` methods, would be nice to add a test

**Review by andrei-epure-sonarsource — CHANGES_REQUESTED** [review]
left some comments

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/ImplementISerializableCorrectly.CSharp9.cs (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/ImplementISerializableCorrectly.CSharp9.cs:18]: It's inherited below. Moved together.

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/ImplementISerializableCorrectly.CSharp9.cs [hunk 5]
```diff
@@ -0,0 +1,282 @@
+﻿using System;
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/ImplementISerializableCorrectly.CSharp9.cs:1]: Added reproducer for #1211

**Review by andrei-epure-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**